### PR TITLE
RANCHER-1700: Migrate deployModuleFromFeatureBranch Pipeline to confirmed Base Class Hierarchy

### PR DESCRIPTION
**Pull request by yaroslavishchenko**
Related Jira Ticket:
[RANCHER-1700:Migrate deployModuleFromFeatureBranch Pipeline to confirmed Base Class Hierarchy](https://folio-org.atlassian.net/browse/RANCHER-1700)

Working Jenkins Job:
https://jenkins-aws.indexdata.com/job/folioRancher/job/tmpFolderForDraftPipelines/job/YaroslavI/job/deployEurekaModuleFromFeatureBranch/268/console

### CODE REVIEW

File: pipelines/folioRancher/folioDevTools/moduleDeployment/deployModuleFromFeatureBranchEureka/Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+ansiColor('xterm') {
+  node(params.AGENT) {
+    try {
+      stage('Ini') {
+        buildName "#${params.MODULE_NAME}.${env.BUILD_ID}"
+        buildDescription "Env: ${namespace.getWorkspaceName()}\nBranch: ${params.MODULE_BRANCH}\n" + "Config: ${params.CONFIG_TYPE}\n${currentBuild.getBuildCauses()[0].shortDescription}"
+      }
+
+      stage('[Git] Checkout module source') {
+        if(skipStage) {
```
Comment (OHaimanov) [inline pipelines/folioRancher/folioDevTools/moduleDeployment/deployModuleFromFeatureBranchEureka/Jenkinsfile:79]: Is this is for testing purposes? And if yes, please remove

File: pipelines/folioRancher/folioDevTools/moduleDeployment/deployModuleFromFeatureBranchEureka/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        module.loadModuleDetails('mod-lists-2.1.0-SNAPSHOT.95')
+      }
+
+      // 1. Register Application Descriptor with Updated Module Version
+      stage('[Rest] Update App Descriptor') {
+//        if(skipStage) {
+//          logger.info("Skipping stage")
+//          return
+//        }
+        eureka.updateAppDescriptorFlow(namespace.applications, module)
```
Comment (OHaimanov) [inline pipelines/folioRancher/folioDevTools/moduleDeployment/deployModuleFromFeatureBranchEureka/Jenkinsfile:157]: Did I miss update of application descriptor with new module version?

File: pipelines/folioRancher/folioDevTools/moduleDeployment/deployModuleFromFeatureBranchEureka/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+      }
+
+      stage('Generate context') {
+        namespace.tenants = eureka.getExistedTenantsFlow()
+        logger.debug("Configured Tenants: ${namespace.tenants}")
+
+        namespace.withApplications(eureka.getEnabledApplicationsFlow(namespace.tenants))
+        logger.debug("Configured Applications: ${namespace.applications}")
+
+        // TODO: debug code to be removed soon
```
Comment (OHaimanov) [inline pipelines/folioRancher/folioDevTools/moduleDeployment/deployModuleFromFeatureBranchEureka/Jenkinsfile:147]: Please do not forget to remove this one

File: src/org/folio/rest_v2/eureka/Eureka.groovy [hunk 4]
```diff
@@ -212,4 +215,139 @@ class Eureka extends Base {
 
     return this
   }
+
+  /**
+   * Get Configured Tenants on Environment Namespace.
+   * @return Map of EurekaTenant objects.
+   */
+  Map<String, EurekaTenant> getExistedTenantsFlow() {
```
Comment (OHaimanov) [inline src/org/folio/rest_v2/eureka/Eureka.groovy:223]: Flow is something that doesn't return anything or just metadata
I think for this method will be good to leave simply getExistedTenants()

File: src/org/folio/rest_v2/eureka/Eureka.groovy [hunk 5]
```diff
@@ -212,4 +215,139 @@ class Eureka extends Base {
...
+    }
+
+    return configuredTenantsMap
+  }
+
+  /**
+   * Get Existed Applications on Environment Namespace.
+   * @return Map of Application Name and Application ID.
+   */
+  static Map<String, String> getEnabledApplicationsFlow(Map<String, EurekaTenant> tenants) {
```
Comment (OHaimanov) [inline src/org/folio/rest_v2/eureka/Eureka.groovy:251]: Same here

File: src/org/folio/rest_v2/eureka/Eureka.groovy [hunk 6]
```diff
@@ -212,4 +215,139 @@ class Eureka extends Base {
...
+    // Get Application Descriptor Updated with New Module Version
+    appDescriptorsMap.each { appId, descriptor->
+      // Get Incremental Number for New Application Version
+      String incrementalNumber = descriptor['version'].toString().tokenize('.').last().toInteger() + 1
+
+      // Update existing Application Descriptor with New Module Version
+      Map updatedAppDescriptor = getUpdatedApplicationDescriptor(descriptor as Map, module, incrementalNumber)
+
+      // Register Updated Application Descriptor to Environment
+      // TODO: uncomment following line after testing
```
Comment (OHaimanov) [inline src/org/folio/rest_v2/eureka/Eureka.groovy:292]: Is testing complete?

File: pipelines/folioRancher/folioDevTools/moduleDeployment/deployModuleFromFeatureBranchEureka/Jenkinsfile (hunk 1 above)
Comment (yaroslavishchenko) [reply pipelines/folioRancher/folioDevTools/moduleDeployment/deployModuleFromFeatureBranchEureka/Jenkinsfile:79]: @OHaimanov 
Yep you're right its just to speed up testing process.
Once evrything works and looks good these lines will be deleted.

File: pipelines/folioRancher/folioDevTools/moduleDeployment/deployModuleFromFeatureBranchEureka/Jenkinsfile (hunk 2 above)
Comment (yaroslavishchenko) [reply pipelines/folioRancher/folioDevTools/moduleDeployment/deployModuleFromFeatureBranchEureka/Jenkinsfile:157]: @OHaimanov 
Application descriptor update with new module version is performed inside of **updateAppDescriptorFlow**.

File: pipelines/folioRancher/folioDevTools/moduleDeployment/deployModuleFromFeatureBranchEureka/Jenkinsfile (hunk 3 above)
Comment (yaroslavishchenko) [reply pipelines/folioRancher/folioDevTools/moduleDeployment/deployModuleFromFeatureBranchEureka/Jenkinsfile:147]: Thanks. 
Thats why there is such comment with TODO.

File: src/org/folio/rest_v2/eureka/Eureka.groovy (hunk 6 above)
Comment (yaroslavishchenko) [reply src/org/folio/rest_v2/eureka/Eureka.groovy:292]: **folio-eureka-volya** environment is a bit compromised.
So there is deploed very fresh **folio-tmp-etest01** environment.
Today this line will be re-tested in scope of end to end pipeline testing.

File: src/org/folio/rest_v2/eureka/Eureka.groovy (hunk 4 above)
Comment (yaroslavishchenko) [reply src/org/folio/rest_v2/eureka/Eureka.groovy]: Enhanced accordingly to your comment.

File: src/org/folio/rest_v2/eureka/Eureka.groovy (hunk 5 above)
Comment (yaroslavishchenko) [reply src/org/folio/rest_v2/eureka/Eureka.groovy]: Enhanced accordingly to your comment.

**Comment by eldiiar-duishenaliev** [comment]
Hello, 
Unfortunately, in overall this PR is still not ready to be merged into our stable branch.
Reasons:
1. There is not logic for handling special case for mod-consortia-keycloak.
2. Newly generated app should be entitled ONLY on top of requested tenant, not all tenants available on env.
3. As was mentioned on one of daily meeting, if entitlement has failed, it should be cleaning up of pre-entitlement resources (registered apps, modules discovery and so on)
4. Additional improvements are required in the logic for updating apps.

**Comment by yaroslavishchenko** [comment]
This pull request is closed in favor of #645 .

**Pull request closed without merging**